File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/MouseCommand.kt
package mono.state.command.mouse

import mono.common.MouseCursor
import mono.graphics.geo.MousePointer
import mono.state.MainStateManager
import mono.state.command.CommandEnvironment

/**
 * A strategy interface for mouse interaction command happens on [MainStateManager]
 */
internal sealed interface MouseCommand {
    /**
     * CSS mouse cursor value to show during command execution.
     */
    val mouseCursor: MouseCursor?

    /**
     * Handles mouse events.
     * Returns true when the action finishes.
     */
    fun execute(environment: CommandEnvironment, mousePointer: MousePointer): CommandResultType

    enum class CommandResultType {
        WORKING,
        WORKING_PHASE2,
        DONE,
        UNKNOWN
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/mouse-command.ts
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import type { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import type { MousePointer } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A strategy interface for mouse interaction command happens on [MainStateManager]
 */
export interface MouseCommand {
    /**
     * CSS mouse cursor value to show during command execution.
     */
    mouseCursor?: MouseCursor;

    execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer,
    ): MouseCommand.CommandResultType;
}


export namespace MouseCommand {
    export enum CommandResultType {
        WORKING,
        WORKING_PHASE2,
        DONE,
        UNKNOWN
    }
}
